tls/s2n_x509_validator.c

```diff
@@ -513,21 +579,31 @@ static S2N_RESULT s2n_x509_validator_verify_cert_chain(struct s2n_x509_validator
                 S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
     }
 
-    uint64_t current_sys_time = 0;
-    RESULT_GUARD(s2n_config_wall_clock(conn->config, &current_sys_time));
-    if (sizeof(time_t) == 4) {
-        /* cast value to uint64_t to prevent overflow errors */
-        RESULT_ENSURE_LTE(current_sys_time, (uint64_t) MAX_32_TIMESTAMP_NANOS);
-    }
+    /* Disabling time validation may set a NO_CHECK_TIME flag on the X509_STORE_CTX. Calling
+     * X509_STORE_CTX_set_time will override this flag. To prevent this, X509_STORE_CTX_set_time is
+     * only called if time validation is enabled.
+     */
```

Comment:
Does this work if you call X509_STORE_CTX_set_time before setting the NO_CHECK_TIME flag instead of after? And would that simplify your testing at all?

Reply:
Unfortunately I don't think the order makes a difference. `X509_STORE_CTX_set_time()` doesn't set the same `NO_CHECK_TIME` flag. Instead, it sets its own flag: `USE_CHECK_TIME`. With some libcryptos, the `USE_CHECK_TIME` flag is preferred, so if `X509_STORE_CTX_set_time()` is called, it doesn't matter that you also set the `NO_CHECK_TIME` flag. This is the logic in OpenSSL for example:
https://github.com/openssl/openssl/blob/025535ecd11bdebd8eb28ed4f0f6b509b1b54577/crypto/x509/x509_vfy.c#L1788-L1793